### DISCUSSION TITLE
Directive for HTML file to be injected into slide(s)

### DISCUSSION THREAD

**Discussion by inform880** [marp-team/marp#226] [Ideas] [answered]
Hi,

I'm really enjoying Marp so far!

In my free time, I'm slowly putting together a slide deck/CSS theme in Marp based off of a PowerPoint presentation for a rather large company. I'm finding that some of our background graphics are complicated and required me to use HTML for setting positions in order for it to look _exactly_ like the PowerPoint slides. I see three options for me in my situation:

1. Render the backgrounds myself and use them as full screen backgrounds for each variant. (Annoying and complicated, and I lose the fact that some of the graphics are in SVG format)
2. Continue what I'm doing and require people making new slides to copy and paste the HTML for the backgrounds in each of their presentations/slides (which would probably kill the idea of anyone using it).
3. **And here's my idea:** a directive that people could use, like the ```class``` directive, that would be a path to an HTML file, that would be injected into the slides, like this:

```md
<!-- presentation.md -->
...
<!-- _html: title-slide -->

# Whatever

slide one
...
```
and
```html
<!-- title-slide.html -->
<img src="assets/top-right-background.png" style="width: 3.75in; top: 0; right: 0; position: absolute;"/>
<img src="assets/bottom-right-background.png" style="width: 3.75in; bottom: 0; right: 0; position: absolute;"/>
<img src="assets/logo-white.png" style="width: 1.2in; top: 0.2in; left: 0.37in; position: absolute;"/>
```
would render something like this(?)
```html
<section>
  <img src="assets/top-right-background.png" style="width: 3.75in; top: 0; right: 0; position: absolute;"/>
  <img src="assets/bottom-right-background.png" style="width: 3.75in; bottom: 0; right: 0; position: absolute;"/>
  <img src="assets/logo-white.png" style="width: 1.2in; top: 0.2in; left: 0.37in; position: absolute;"/>
  <h1>Whatever</h1>
  <p>slide one</p>
</section>
```
Maybe this is a dumb idea for whatever reason, or maybe there's already a way to do this? Anyway, while I'm waiting for feedback on this, I'll be taking a look [here](https://marpit.marp.app/directives?id=custom-directives).

**Comment by yhatt** [comment, marked as answer]
It looks to me like CSS is enough expressive to get to reproduce the layout of background images you've described.

```css
section.title-slide {
  background:
    url('assets/top-right-background.png') no-repeat top right / 3.75in auto,
    url('assets/bottom-right-background.png') no-repeat bottom right / 3.75in auto,
    url('assets/logo-white.png') no-repeat top 0.2in left 0.37in / 1.2in auto;
}
```

```markdown
<!-- _class: title-slide -->

# Whatever

slide one
```

We don't want to make Marp over-complicated so there are no plans like `html` directive, but you can try to add the custom behavior by making a Marp plugin by defining a custom directive.

```javascript
// engine.js
const htmlDirectivePlugin = (md) => {
  // Define "html" local directive
  md.marpit.customDirectives.local.html = (value) => ({ html: value })

  // Override markdown-it renderer for the open tag of Marpit slide
  // https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md#renderer
  const defaultRenderer = (tokens, idx, options, env, self) =>
    self.renderToken(tokens, idx, options)

  const slideOpenRenderer =
    md.renderer.rules.marpit_slide_open || defaultRenderer

  md.renderer.rules.marpit_slide_open = (tokens, idx, options, env, self) => {
    const openTag = slideOpenRenderer(tokens, idx, options, env, self)

    const { marpitDirectives } = tokens[idx].meta
    switch (marpitDirectives && marpitDirectives.html) {
      case 'title-slide':
        return `${openTag}
          <img src="assets/top-right-background.png" style="width: 3.75in; top: 0; right: 0; position: absolute;"/>
          <img src="assets/bottom-right-background.png" style="width: 3.75in; bottom: 0; right: 0; position: absolute;"/>
          <img src="assets/logo-white.png" style="width: 1.2in; top: 0.2in; left: 0.37in; position: absolute;"/>
        `
      default:
        return openTag
    }
  }
}

// Export as a engine for Marp CLI
module.exports = ({ marp }) => marp.use(htmlDirectivePlugin)
```

```console
marp --engine ./engine.js your-markdown.md
```
